drivers/console/display_console.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2019 Intel Corp.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_STDOUT_CONSOLE)
+
+#include <device.h>
+#include <display.h>
+#include <init.h>
+#include <string.h>
+
+static struct device *display;		/* display device */
+static int size_x, size_y;		/* dimensions */
+static int current_x, current_y;	/* cursor position */
+
+/*
+ * For now, we only support the Spleen 8x16 font in lib/fonts/spleen_8x16.c.
+ * It's simple enough to make this more flexible if/when the need arises.
+ */
+
+#define FONT_WIDTH	8
+#define FONT_HEIGHT	16
+#define FONT_PITCH	8
+#define FONT_SIZE	16
+#define FONT_FIRST	32
+#define FONT_LAST	126
+
+extern u8_t font_spleen_8x16[];
+
+const static struct display_buffer_descriptor char_desc = {
+	.width = FONT_WIDTH,
+	.height = FONT_HEIGHT,
+	.pitch = FONT_PITCH,
+	.buf_size = FONT_SIZE
+};
+
+/* Why aren't these hooks declared in a header somewhere? */
+
+extern void __stdout_hook_install(int (*)(int));
+extern void __printk_hook_install(int (*)(int));
+
+/**
+ * @brief Output one character to display console
+ * @param c Character to output
+ * @return The character passed as input.
+ */
+static int display_putchar(int c)
+{
+	switch (c) {
+	case '\r':
+		current_x = 0;
+		break;
+	case '\n':
+		current_x = 0;
+		++current_y;
+		break;
+	default:
+		if ((c >= FONT_FIRST) && (c <= FONT_LAST)) {
+			display_write(display,
+				      current_x * FONT_WIDTH,
+				      current_y * FONT_HEIGHT,
+				      &char_desc,
+				      font_spleen_8x16
+					+ (c - FONT_FIRST) * FONT_SIZE);
+			++current_x;
+		}
+	}
+
+	if (current_x == size_x) { /* wrap */
+		current_x = 0;
+		++current_y;
+	}
+
+	if (current_y == size_y) { /* scroll */
+		int x, y;
+		u8_t buffer[FONT_SIZE];
+
+		for (y = 1; y < size_y; ++y) {
+			for (x = 0; x < size_x; ++x) {
+				display_read(display, x * FONT_WIDTH,
```

Comment:
Would it not be better to put a wrapper around CFB (subsys/fb/cfb.c) and extend it where necessary?

Reply:
I don't believe so.  `display_console.c` allows the use of any display for console output.  `subsys/fb` is specific to framebuffer displays, only supports very specific framebuffer configurations (vertically tiled), and appears to suffer from a fair amount of bitrot.